stdlib/in_channel.mli

```diff
@@ -0,0 +1,148 @@
+    function. *)
+
+val seek : t -> int64 -> unit
+(** [seek chan pos] sets the current reading position to [pos] for channel
+    [chan]. This works only for regular files. On files of other kinds, the
+    behavior is unspecified. *)
+
+val pos : t -> int64
+(** Return the current reading position for the given channel.  For files opened
+    in text mode under Windows, the returned position is approximate (owing to
+    end-of-line conversion); in particular, saving the current position with
+    {!pos}, then going back to this position using {!seek} will not work.  For
+    this programming idiom to work reliably and portably, the file must be
+    opened in binary mode. *)
+
+val length : t -> int64
+(** Return the size (number of characters) of the regular file on which the
+    given channel is opened.  If the channel is opened on a file that is not a
+    regular file, the result is meaningless.  The returned size does not take
+    into account the end-of-line translations that can be performed when reading
+    from a channel opened in text mode. *)
+
+val close : t -> unit
+(** Close the given channel.  Input functions raise a [Sys_error] exception when
+    they are applied to a closed input channel, except {!close}, which does
+    nothing when applied to an already closed channel. *)
+
+val close_noerr : t -> unit
+(** Same as {!close}, but ignore all errors. *)
+
+val input_char : t -> char option
+(** Read one character from the given input channel.  Returns [None] if there
+    are no more characters to read. *)
+
+val input_byte : t -> int option
+(** Same as {!input_char}, but return the 8-bit integer representing the
+    character. Returns [None] if the end of file was reached. *)
+
+val input_line : t -> string option
+(** [input_line ic] reads characters from [ic] until a newline or the end of
+    file is reached.  Returns the string of all characters read, without the
+    newline (if any).  Returns [None] if the end of the file has been reached.
+    In particular, this will be the case if the last line of input is empty.
+
+    A newline is the character [\n] unless the file is open in text mode and
+    {!Sys.win32} is [true] in which case it is the sequence of characters
+    [\r\n]. *)
+
+val input_binary_int : t -> int option
+(** Read an integer encoded in binary format (4 bytes, big-endian) from the
+    given input channel. See {!Out_channel.output_binary_int}.  Returns [None]
+    if the end of file was reached while reading the integer. *)
+
+val input : t -> bytes -> int -> int -> int
+(** [input ic buf pos len] reads up to [len] characters from the given channel
+    [ic], storing them in byte sequence [buf], starting at character number
+    [pos].  It returns the actual number of characters read, between 0 and [len]
+    (inclusive).
+
+    A return value of 0 means that the end of file was reached.
+
+    A return value between 0 and [len] exclusive means that not all requested
```

Comment:
```suggestion
    A return value between 1 and [len] exclusive means that not all requested
```

With the current definition, it's not possible to reliably distinguish an EOF from a short return of 0.  The [read(2)](https://pubs.opengroup.org/onlinepubs/009696899/functions/read.html) definition does guarantee that for the case of regular files that 0 indicates an eof, but doesn't entirely guarantee that a 0 _could_ be returned to indicate an unfortunately timed signal. 

But given that pretty much every use of `input` in OCaml code assumes that `0` indicates an EOF, should we just define it as such and treat a non-eof zero return as an implementation bug that should be fixed in the OCaml runtime?  This would only occur with blocking sockets, as non-blocking ones would signal an EAGAIN or EINTR in any reasonable OS.

Reply:
> The [read(2)](https://pubs.opengroup.org/onlinepubs/009696899/functions/read.html) definition does guarantee that for the case of regular files that 0 indicates an eof, but doesn't entirely guarantee that a 0 _could_ be returned to indicate an unfortunately timed signal.

Hi Anil, I have not looked at these intricacies before, but a brief look at `osdeps.h` discovered the following comment:
https://github.com/ocaml/ocaml/blob/f0d61fa106f9cb6a11a36852ca44cf7d33d7ca10/runtime/caml/osdeps.h#L43-L45
(here `caml_read_fd` is the function that implements `input`). The special return code `Io_interrupted` is handled by simply retrying in the read loop:
https://github.com/ocaml/ocaml/blob/f0d61fa106f9cb6a11a36852ca44cf7d33d7ca10/runtime/io.c#L790-L808

Does this make sense, or am I missing something?

Reply:
Thanks for the pointer @nojb. I think the existing code there shouldn't return a 0 unless it's an EOF if the underlying descriptor is blocking (which it should be, since the runtime initialises it as such). So clarifying the definition to say that `0` is always EOF seems ok to me.

(@Engil @kayceesrk I'm just double-checking what multicore OCaml does with multiple domains, since concurrent access to an underlying non-regular-file fd isn't well defined. We reuse the systhreads mutex locking machinery right now so the semantics shouldn't change there except for an edge case post-fork in https://github.com/ocaml-multicore/ocaml-multicore/issues/614. Apologies for the noise for those with single cores ;-)

Reply:
> if the underlying descriptor is blocking (which it should be, since the runtime initialises it as such).

But there is also `Unix.{in_out}_channel_of_descr`. There's a check [there](https://github.com/ocaml/ocaml/blob/f0d61fa106f9cb6a11a36852ca44cf7d33d7ca10/otherlibs/unix/channels.c#L31-L65) to rule out some fd types. As far as I can tell it doesn't reject fds set to non-blocking mode. 

Shouldn't it ?  

`{In,Out}_channel.t` are definitively synchronous abstractions, if they start returning EOF to say wait and try again, then we get into separation of concerns disaster (and as @avsm mentioned earlier likely no ocaml code out there is able to deal with that).

Reply:
> Shouldn't it ?

I think so -- a PR to discuss that would be welcome (I have the feeling it may not be entirely trivial to tell whether a file descriptor is blocking or not).

Reply:
> So clarifying the definition to say that `0` is always EOF seems ok to me.

Are you suggesting to commit your patch (replacing `0` by `1`)? I don't have a particular problem doing so, but I always understood the sentence `between 0 and [len] exclusive` as meaning the interval `[1, len-1]` (ie the `exclusive` being applied to both ends of the interval), especially since the sentence just before says explicitly that a return value of `0` means the end of file has been reached.

I want to make sure there is some consensus around the meaning of the words before making any change.

If we *do* go ahead with the change, we should also adjust the docstring in `Stdlib`.

Reply:
> I think so -- a PR to discuss that would be welcome (I have the feeling it may not be entirely trivial to tell whether a file descriptor is blocking or not).

For Unix I think it should be a matter of calling [`fcntl`](https://man7.org/linux/man-pages/man2/fcntl.2.html) with `F_GETFL` and check for `O_NONBLOCK` flag (don't know about the Windows part). But maybe there's a reason why it doesn't do so.

Reply:
> but I always understood the sentence `between 0 and [len] exclusive` as meaning the interval `[1, len-1]`

I know people like their exclusive bounds but writing the interval explicitely is just clearer in my opinion. I certainly got it wrong. 

But if you ask me, I think the whole thing is needlessly verbose. It's already mentioned above that less than `len` byte can be read. I would just cut the whole thing to: 

```
[input ic buf pos len] reads up to [len] characters from the given 
channel [ic], storing them in byte sequence [buf], starting at character 
number [pos]. It returns the actual number of characters read, between 0 
and [len] (inclusive). A return value of 0 means that the end of file 
was reached.

Use {!really_input} to read exactly [len] characters. 

@raise Invalid_argument if [pos] and [len] do not designate a valid range 
of [buf].
```

Reply:
I don't want to get bogged down on this minor point, so if we can't converge quickly I would suggest to open a different PR to discuss it.

@dbuenzli The existing docstring contains a pointer about calling `input` again if less than `len` characters have been read which I feel is useful for beginners. Can we put this fragment back in the docstring while at the same time clarifying the different return values?

Reply:
I don't feel it's necessary to state the obvious. If you really do you I'd suggest to add this sentence to the first paragraph of my proposal:`If less than [len] characters were read and the end of file is not reached, {!input} can be called again to read the remaining characters.`

Reply:
Thanks, after some further thought I went with the earlier, more succint, version, which I committed.